src-rocksdb/RocksDB.hs

```diff
@@ -0,0 +1,313 @@
+    rocksdb_write dbPtr opts batch errPtr
+    assertErrPtr "rocksdb_write" errPtr
+
+checkpoint :: RocksDB -> FilePath -> IO ()
+checkpoint (RocksDB dbPtr errPtr) path =
+    withCString path $ \path' ->
+    bracket rocksdb_checkpoint_object_create' rocksdb_checkpoint_object_destroy $ \cp -> do
+        rocksdb_checkpoint_create cp path' 0 errPtr
+        assertErrPtr "rocksdb_checkpoint_create" errPtr
+  where
+    rocksdb_checkpoint_object_create' = do
+        cp <- rocksdb_checkpoint_object_create dbPtr errPtr
+        assertErrPtr "rocksdb_checkpoint_object_create" errPtr
+        return cp
+
+-------------------------------------------------------------------------------
+-- write batch
+-------------------------------------------------------------------------------
+
+newtype WriteBatch = WriteBatch (Ptr WRITEBATCH)
+
+withWriteBatch :: (WriteBatch -> IO r) -> IO r
+withWriteBatch kont = bracket
+    rocksdb_writebatch_create
+    rocksdb_writebatch_destroy
+    (\ptr -> kont (WriteBatch ptr))
+
+writeBatchPut :: WriteBatch -> ByteString -> ByteString -> IO ()
+writeBatchPut (WriteBatch batch) key val =
+    unsafeUseAsCStringLen key $ \(kp, kl) ->
+    unsafeUseAsCStringLen val $ \(vp, vl) ->
+        rocksdb_writebatch_put batch kp (intToCSize kl) vp (intToCSize vl)
+
+writeBatchDelete :: WriteBatch -> ByteString -> IO ()
+writeBatchDelete (WriteBatch batch) key =
+    unsafeUseAsCStringLen key $ \(kp, kl) ->
+        rocksdb_writebatch_delete batch kp (intToCSize kl)
+
+-------------------------------------------------------------------------------
+-- block based table options
+-------------------------------------------------------------------------------
+
+newtype BlockTableOptions = BlockTableOptions (Ptr BLOCKTABLEOPTIONS)
+
+withBlockTableOptions :: (BlockTableOptions -> IO r) -> IO r
+withBlockTableOptions kont = bracket
+    rocksdb_block_based_options_create
+    rocksdb_block_based_options_destroy
+    (\ptr -> kont (BlockTableOptions ptr))
+
+blockBasedOptionsSetFilterPolicy :: BlockTableOptions -> FilterPolicy -> IO ()
+blockBasedOptionsSetFilterPolicy (BlockTableOptions opts) (FilterPolicy policy) =
+    rocksdb_block_based_options_set_filter_policy opts policy
+
+optionsSetBlockBasedTableFactory :: Options -> BlockTableOptions -> IO ()
+optionsSetBlockBasedTableFactory (Options opts) (BlockTableOptions ptr) =
+    rocksdb_options_set_block_based_table_factory opts ptr
+
+-------------------------------------------------------------------------------
+-- filter policy
+-------------------------------------------------------------------------------
+
+newtype FilterPolicy = FilterPolicy (Ptr FILTERPOLICY)
+
+withFilterPolicyBloom :: Int -> (FilterPolicy -> IO r) -> IO r
+withFilterPolicyBloom bits_per_key kont = bracket
+    (rocksdb_filterpolicy_create_bloom (intToCInt bits_per_key))
+    (\_ -> return ()) -- rocksdb_filterpolicy_destroy, causes segfault if we free it.
```

Comment:
So what is the consequence? Is it a problem that we don't free it?

Reply:
> Is it a problem that we don't free it?

The ownerships are not clear from the docs (the docs are essentially only a header file without much comments). Maybe nothing leaks. (And even it it does, the leak is quite minimal; this is not in a loop).